src/resolver.ts

```diff
@@ -3130,12 +3138,83 @@ export class Resolver extends DiagnosticEmitter {
+                +==================+=========+===========+=========+
+                | Visibility Table | Private | Protected | Public  |
+                +==================+=========+===========+=========+
+                | Private          | error   | error     | error   |
+                +------------------+---------+-----------+---------+
+                | Protected        | error   | allowed   | error   |
+                +------------------+---------+-----------+---------+
+                | Public           | error   | allowed   | allowed |
+                +------------------+---------+-----------+---------+
+              */
+
+              let baseClass = <Class>base;
+
+              // handle cases row-by-row
+              if (fieldPrototype.is(CommonFlags.PRIVATE)) {
+                if (existingField.is(CommonFlags.PRIVATE)) {
+                  this.errorRelated(
+                    DiagnosticCode.Types_have_separate_declarations_of_a_private_property_0,
+                    fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                    fieldPrototype.name
+                  );
+                } else {
+                  this.errorRelated(
+                    DiagnosticCode.Property_0_is_private_in_type_1_but_not_in_type_2,
+                    fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                    fieldPrototype.name, instance.internalName, baseClass.internalName
+                  );
+                }
+              } else if (fieldPrototype.is(CommonFlags.PROTECTED)) {
+                if (existingField.is(CommonFlags.PRIVATE)) {
+                  this.errorRelated(
+                    DiagnosticCode.Property_0_is_private_in_type_1_but_not_in_type_2,
+                    fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                    fieldPrototype.name, baseClass.internalName, instance.internalName
+                  );
+                } else if (!existingField.is(CommonFlags.PROTECTED)) {
+                  // may be implicitly public
+                  this.errorRelated(
+                    DiagnosticCode.Property_0_is_protected_in_type_1_but_public_in_type_2,
+                    fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                    fieldPrototype.name, instance.internalName, baseClass.internalName
+                  );
+                }
+              } else {
+                // fieldPrototype is public here
+                if (existingField.is(CommonFlags.PRIVATE)) {
+                  this.errorRelated(
+                    DiagnosticCode.Property_0_is_private_in_type_1_but_not_in_type_2,
+                    fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                    fieldPrototype.name, baseClass.internalName, instance.internalName
+                  );
+                }
+              }
```

Comment:
Wondering: Would this become shorter when checking for existingField first, since when it is private, that's a common case for all essentially?

Reply:
The errors are different if both fields are private. I think that would make the checks a bit harder to read.